subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/DefaultJavaToolchainRepositoryRegistry.java

```diff
@@ -0,0 +1,146 @@
+        this.registrationBroadcaster = registrationBroadcaster;
+        this.instantiator = instantiator;
+        this.objectFactory = objectFactory;
+        this.providerFactory = providerFactory;
+        this.authenticationSchemeRegistry = authenticationSchemeRegistry;
+    }
+
+    @Override
+    public <T extends JavaToolchainRepository> void register(String name, Class<T> implementationType) {
+        if (registrations.containsKey(name)) {
+            throw new GradleException("Duplicate " + JavaToolchainRepository.class.getSimpleName() + " registration under the name '" + name + "'");
+        }
+
+        Provider<T> provider = sharedServices.registerIfAbsent(name, implementationType, EMPTY_CONFIGURE_ACTION);
+        JavaToolchainRepositoryRegistrationInternal registration = new DefaultJavaToolchainRepositoryRegistration(name, provider);
+        registrations.put(name, registration);
+
+        registrationBroadcaster.onRegister(registration);
+    }
+
+    @Override
+    public void request(String registrationName) {
+        request(findRegistrationByName(registrationName));
+    }
+
+    @Override
+    public void request(String registrationName, Action<? super AuthenticationSupported> authentication) {
+        request(findRegistrationByName(registrationName), authentication);
+    }
+
+    @Override
+    public void request(JavaToolchainRepositoryRegistration registration) {
+        requests.add(createRequest((JavaToolchainRepositoryRegistrationInternal) registration));
+    }
+
+    @Override
+    public void request(JavaToolchainRepositoryRegistration registration, Action<? super AuthenticationSupported> authentication) {
+        JavaToolchainRepositoryRequest request = createRequest((JavaToolchainRepositoryRegistrationInternal) registration);
+        authentication.execute(request);
+        requests.add(request);
+    }
+
+    private JavaToolchainRepositoryRequest createRequest(JavaToolchainRepositoryRegistrationInternal registration) {
+        DefaultAuthenticationContainer authenticationContainer = new DefaultAuthenticationContainer(instantiator, CollectionCallbackActionDecorator.NOOP);
+        for (Map.Entry<Class<Authentication>, Class<? extends Authentication>> e : authenticationSchemeRegistry.getRegisteredSchemes().entrySet()) {
+            authenticationContainer.registerBinding(e.getKey(), e.getValue());
+        }
+
+        AuthenticationSupporter authenticationSupporter = new AuthenticationSupporter(instantiator, objectFactory, authenticationContainer, providerFactory);
+        return objectFactory.newInstance(JavaToolchainRepositoryRequest.class, registration, authenticationSupporter, providerFactory);
+    }
+
+    @Override
+    public List<? extends JavaToolchainRepositoryRegistration> allRequestedRegistrations() {
+        return requests.stream().map(JavaToolchainRepositoryRequest::getRegistration).collect(Collectors.toList()); //TODO (#21082): do we need to create a defensive copy here?
```

Comment:
Since these can get exposed in the DSL, they should be immutable at least.
The question is whether we want to expose a live version. That is doing a get, then registering, is that registration reflected in the previous get result?